File: kattis/joinstrings.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		Kattio io = new Kattio(System.in);
		int n = io.getInt();
		int inf = 1000000000;
		String[] arr = new String[100010];
		int[] nodelst = new int[100010];
		boolean[] boolarr = new boolean[100010];
		HashMap<Integer, Integer> al = new HashMap<>();

		for (int i = 1; i <= n; i++) {
			String c = io.getWord();
			arr[i] = c;
			nodelst[i] = i;
			al.put(i, inf);
			boolarr[i] = true;
		}

		for (int i = 1; i <= n - 1; i++) {
			int a = io.getInt();
			int b = io.getInt();
			int temp = nodelst[a];
			nodelst[a] = nodelst[b];
			boolarr[b] = false;
			al.put(temp, b);
		}

		int k = -1;
		for (int i = 1; i <= n; i++) {
			if (boolarr[i]) {
				k = i;
				break;
			}
		}

		while (true) {
			io.print(arr[k]);
			if (al.get(k) == inf) {
				break;
			}
			else {
				k = al.get(k);
			}
		}

		io.print('\n');
		io.flush();
	}
}

File: kattis/joinstrings.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	int n;
	cin >> n;
	int inf = 1e9;
	string arr[100010];
	int nodelst[100010];
	bool boolarr[100010];
	unordered_map<int, int> al;
	for (int i = 1; i <= n; i++) {
		string c;
		cin >> c;
		arr[i] = c;
		nodelst[i] = i;
		al[i] = inf;
		boolarr[i] = true;

	}
	for (int i = 1; i <= n-1; i++) {
		int a, b;
		cin >> a >> b;
		int temp = nodelst[a];
		nodelst[a] = nodelst[b];
		boolarr[b] = false;
		al[temp] = b;
	}
	int k;
	for (int i = 1; i <= n; i++) {
		if (boolarr[i]) {
			k = i;
			break;
		}
	}
	while (true) {
		cout << arr[k];
		if (al[k] == inf) {
			break;
		}
		else {
			k = al[k];
		}
	}
	cout << endl;
}